url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/chunk/RIFFChunk.java

```diff
@@ -0,0 +1,247 @@
+
+	public RIFFChunk(KlangFile parentFile, long startpos, long endpos, long datastartpos, long dataendpos, int chunkType, String chunkName, ContainerChunk parentChunk, Boolean isBigEndian, RandomAccessFile raf) throws IOException {
+		super(parentFile, startpos, endpos, datastartpos, dataendpos, chunkType, chunkName, parentChunk, isBigEndian, raf);
+		this.usesPadByte = true;
+		raf.seek(datastartpos);
+		/*
+		 * LOAD PRIMITIVE OBJECTS (UNINITIALIZED)
+		 */
+		FourCC name = new FourCC();
+		PrimitiveData nameData;
+		/*
+		 * READ FROM FILE TO FILL PRIMITIVE DATA
+		 */
+		if (endpos >= raf.getFilePointer() + 4) {
+			try {
+				long bytepos = raf.getFilePointer();
+				name.setValueFromFile(raf);
+				nameData = new PrimitiveData(name, KlangConstants.PRIMITIVE_DATA_RIFF_CHUNK_NAME, bytepos, raf.getFilePointer());
+			} catch (BadValueException er) {
+				nameData = new PrimitiveData(name, KlangConstants.PRIMITIVE_DATA_RIFF_CHUNK_NAME, null, null);
+			}
+		} else {
+			nameData = new PrimitiveData(name, KlangConstants.PRIMITIVE_DATA_RIFF_CHUNK_NAME, null, null);
+		}
+		/*
+		 * ADD PRIMITIVES TO VECTOR
+		 */
+		this.primitives.add(nameData);
+		/*
+		 * ADD SUBCHUNKS
+		 */
+		raf.seek(datastartpos + 4);
+		long pointer = raf.getFilePointer();
+		while (pointer < dataendpos) {
+    		Chunk cChunk = ChunkFactory.createChunk(parentFile, pointer, KlangConstants.CHUNKTYPE_RIFF, this, raf, isBigEndian, null, null);
+    		this.chunks.add(cChunk);
+    		pointer = cChunk.getEndPosition();
+    	}
+	}
+	
+	public void addChunkFromFile(File theFile, String name, int position, RandomAccessFile raff) throws IOException {
+		if (!(this.getParentFile() instanceof EditableFileBase)) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		long newsize = (this.dataEndPosition - this.dataStartPosition) + theFile.length() + 8;
+		boolean addPad = false;
+		if (KlangUtil.adjustForPadByte(theFile.length()) != theFile.length()) {
+			addPad = true;
+			newsize = newsize + 1;
+		}
+		if (newsize < 0) {
+			throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+		}
+		try {
+			//first write new file size
+			Primitive size;
+			if (isBigEndian) {
+				size = new Int4ByteUnsignedBE();
+			} else {
+				size = new Int4ByteUnsignedLE();
+			}
+			size.setValueFromString(Long.toString(newsize));
+			raff.seek(startPosition + 4);
+			size.writeValueToFile(raff);
+			//now warn parent chunks
+			this.chunkIsAboutToChangeSize(newsize, raff);
+			//now make the addition
+			long startingplace = 0;
+			if (position > this.getSubChunks().size() + 2) {
+				position = this.getSubChunks().size() + 1;
+			}
+			if (position == this.getSubChunks().size() + 1) {
+				if (this.getSubChunks().size() == 0) {
+					startingplace = this.endPosition;
+				} else {
+					startingplace = this.getSubChunks().get(position - 2).getEndPosition();
+				}
+			} else {
+				startingplace = this.getSubChunks().get(position - 1).getStartPosition();
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 94:102, 164:172.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.